packages/typespec-azure-portal-core/src/decorators.ts

```diff
@@ -0,0 +1,253 @@
+import { ArmResourceKind, getArmResourceKind } from "@azure-tools/typespec-azure-resource-manager";
+import {
+  StringLiteral,
+  getDirectoryPath,
+  getSourceLocation,
+  normalizePath,
+  resolvePath,
+  validateDecoratorUniqueOnNode,
+  type DecoratorContext,
+  type Model,
+  type ModelProperty,
+  type Program,
+  type Type,
+} from "@typespec/compiler";
+import fs from "fs";
+import { PortalCoreKeys } from "./keys.js";
+import { reportDiagnostic } from "./lib.js";
+import { AboutOptions, BrowseOptions, marketplaceOfferOptions } from "./types.js";
+
+/**
+ * This is a Browse decorator which will be use to put more info on the browse view.
+ * @param target The model that is being decorated.
+ * @param options BrowseOptions of the property.
+ */
+export function $browse(context: DecoratorContext, target: Model, options: BrowseOptions) {
+  const { program } = context;
+  validateDecoratorUniqueOnNode(context, target, $browse);
+  isARMResource(program, target, "browse");
+  if (options && (options as Model).properties) {
+    const query = (options as Model).properties.get("argQuery");
+    if (query && query.type.kind === "Model") {
+      const sourceLocation = getSourceLocation(target);
+      const dirPath =
+        sourceLocation && sourceLocation.file.path && getDirectoryPath(sourceLocation.file.path);
+      let argQueryPath = query.type && (query.type as Model).properties.get("filePath");
```

Comment:
question on this!

Reply:
chat with Will : how to get the sourceLocation of the decorator instead of the target

Reply:
Do you mean you want to emit a diagnostic on the decorator itself instead of the type it decorates?
context.decoratorTarget I think would be that or context.getArgumentAt()

Reply:
@timotheeguerin theres two different things that kind of got mixed up in this thread. What I suggested when chatting with @yejee94 is that we should be doing everything relative to the decorator application instead of the target, both the diagnostics, which is simple through ctx.decoratorTarget, but also computing the relative file path relative to the decorator rather than relative to the target, which could be important in the case of augmentations.